File: dmn-signavio/src/test/resources/dmn2java/exported/proto/expected/proto3/java/dmn/date-time-proto/type/TCompositeDateTime.java
package type;

import java.util.*;

@javax.annotation.Generated(value = {"itemDefinitionInterface.ftl", "tCompositeDateTime"})
@com.fasterxml.jackson.annotation.JsonPropertyOrder(alphabetic = true)
@com.fasterxml.jackson.databind.annotation.JsonDeserialize(as = type.TCompositeDateTimeImpl.class)
public interface TCompositeDateTime extends com.gs.dmn.runtime.DMNType {
    static TCompositeDateTime toTCompositeDateTime(Object other) {
        if (other == null) {
            return null;
        } else if (TCompositeDateTime.class.isAssignableFrom(other.getClass())) {
            return (TCompositeDateTime)other;
        } else if (other instanceof com.gs.dmn.runtime.Context) {
            TCompositeDateTimeImpl result_ = new TCompositeDateTimeImpl();
            result_.setDate((javax.xml.datatype.XMLGregorianCalendar)((com.gs.dmn.runtime.Context)other).get("Date", "Date"));
            result_.setTime((javax.xml.datatype.XMLGregorianCalendar)((com.gs.dmn.runtime.Context)other).get("Time", "Time"));
            result_.setDateTime((javax.xml.datatype.XMLGregorianCalendar)((com.gs.dmn.runtime.Context)other).get("DateTime", "DateTime"));
            return result_;
        } else if (other instanceof com.gs.dmn.runtime.DMNType) {
            return toTCompositeDateTime(((com.gs.dmn.runtime.DMNType)other).toContext());
        } else if (other instanceof proto.TCompositeDateTime) {
            TCompositeDateTimeImpl result_ = new TCompositeDateTimeImpl();
            result_.setDate(com.gs.dmn.signavio.feel.lib.DefaultSignavioLib.INSTANCE.date(((proto.TCompositeDateTime) other).getDate()));
            result_.setTime(com.gs.dmn.signavio.feel.lib.DefaultSignavioLib.INSTANCE.time(((proto.TCompositeDateTime) other).getTime()));
            result_.setDateTime(com.gs.dmn.signavio.feel.lib.DefaultSignavioLib.INSTANCE.dateAndTime(((proto.TCompositeDateTime) other).getDateTime()));
            return result_;
        } else {
            throw new com.gs.dmn.runtime.DMNRuntimeException(String.format("Cannot convert '%s' to '%s'", other.getClass().getSimpleName(), TCompositeDateTime.class.getSimpleName()));
        }
    }

    static proto.TCompositeDateTime toProto(TCompositeDateTime other) {
        proto.TCompositeDateTime.Builder result_ = proto.TCompositeDateTime.newBuilder();
        if (other != null) {
            String dateProto_ = com.gs.dmn.signavio.feel.lib.DefaultSignavioLib.INSTANCE.string(((TCompositeDateTime) other).getDate());
            result_.setDate(dateProto_);
            String timeProto_ = com.gs.dmn.signavio.feel.lib.DefaultSignavioLib.INSTANCE.string(((TCompositeDateTime) other).getTime());
            result_.setTime(timeProto_);
            String dateTimeProto_ = com.gs.dmn.signavio.feel.lib.DefaultSignavioLib.INSTANCE.string(((TCompositeDateTime) other).getDateTime());
            result_.setDateTime(dateTimeProto_);
        }
        return result_.build();
    }

    static List<proto.TCompositeDateTime> toProto(List<TCompositeDateTime> other) {
        if (other == null) {
            return null;
        } else {
            return other.stream().map(o -> toProto(o)).collect(java.util.stream.Collectors.toList());
        }
    }

    @com.fasterxml.jackson.annotation.JsonGetter("Date")
    javax.xml.datatype.XMLGregorianCalendar getDate();

    @com.fasterxml.jackson.annotation.JsonGetter("Time")
    javax.xml.datatype.XMLGregorianCalendar getTime();

    @com.fasterxml.jackson.annotation.JsonGetter("DateTime")
    javax.xml.datatype.XMLGregorianCalendar getDateTime();

    default com.gs.dmn.runtime.Context toContext() {
        com.gs.dmn.runtime.Context context = new com.gs.dmn.runtime.Context();
        context.put("date", getDate());
        context.put("time", getTime());
        context.put("dateTime", getDateTime());
        return context;
    }

    default boolean equalTo(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        TCompositeDateTime other = (TCompositeDateTime) o;
        if (this.getDate() != null ? !this.getDate().equals(other.getDate()) : other.getDate() != null) return false;
        if (this.getDateTime() != null ? !this.getDateTime().equals(other.getDateTime()) : other.getDateTime() != null) return false;
        if (this.getTime() != null ? !this.getTime().equals(other.getTime()) : other.getTime() != null) return false;

        return true;
    }

    default int hash() {
        int result = 0;
        result = 31 * result + (this.getDate() != null ? this.getDate().hashCode() : 0);
        result = 31 * result + (this.getDateTime() != null ? this.getDateTime().hashCode() : 0);
        result = 31 * result + (this.getTime() != null ? this.getTime().hashCode() : 0);
        return result;
    }

    default String asString() {
        StringBuilder result_ = new StringBuilder("{");
        result_.append("Date=" + getDate());
        result_.append(", DateTime=" + getDateTime());
        result_.append(", Time=" + getTime());
        result_.append("}");
        return result_.toString();
    }
}

File: dmn-signavio/src/test/resources/dmn2java/exported/proto/expected/proto3/kotlin/dmn/date-time-proto/type/TCompositeDateTime.kt
package type

import java.util.*

@javax.annotation.Generated(value = ["itemDefinitionInterface.ftl", "tCompositeDateTime"])
@com.fasterxml.jackson.annotation.JsonPropertyOrder(alphabetic = true)
@com.fasterxml.jackson.databind.annotation.JsonDeserialize(`as` = type.TCompositeDateTimeImpl::class)
interface TCompositeDateTime : com.gs.dmn.runtime.DMNType {
    @get:com.fasterxml.jackson.annotation.JsonGetter("Date")
    val date: javax.xml.datatype.XMLGregorianCalendar?

    @get:com.fasterxml.jackson.annotation.JsonGetter("Time")
    val time: javax.xml.datatype.XMLGregorianCalendar?

    @get:com.fasterxml.jackson.annotation.JsonGetter("DateTime")
    val dateTime: javax.xml.datatype.XMLGregorianCalendar?

    override fun toContext(): com.gs.dmn.runtime.Context {
        val context = com.gs.dmn.runtime.Context()
        context.put("date", this.date)
        context.put("time", this.time)
        context.put("dateTime", this.dateTime)
        return context
    }

    fun equalTo(o: Any?): Boolean {
        if (this === o) return true
        if (javaClass != o?.javaClass) return false

        val other = o as TCompositeDateTime
        if (if (this.date != null) this.date != other.date else other.date != null) return false
        if (if (this.dateTime != null) this.dateTime != other.dateTime else other.dateTime != null) return false
        if (if (this.time != null) this.time != other.time else other.time != null) return false

        return true
    }

    fun hash(): Int {
        var result = 0
        result = 31 * result + (if (this.date != null) this.date.hashCode() else 0)
        result = 31 * result + (if (this.dateTime != null) this.dateTime.hashCode() else 0)
        result = 31 * result + (if (this.time != null) this.time.hashCode() else 0)
        return result
    }

    fun asString(): String {
        val result_ = StringBuilder("{")
        result_.append("Date=" + date)
        result_.append(", DateTime=" + dateTime)
        result_.append(", Time=" + time)
        result_.append("}")
        return result_.toString()
    }

    companion object {
        @JvmStatic
        fun toTCompositeDateTime(other: Any?): TCompositeDateTime? {
            if (other == null) {
                return null
            } else if (other is TCompositeDateTime?) {
                return other
            } else if (other is com.gs.dmn.runtime.Context) {
                var result_ = TCompositeDateTimeImpl()
                result_.date = other.get("Date", "Date") as javax.xml.datatype.XMLGregorianCalendar?
                result_.time = other.get("Time", "Time") as javax.xml.datatype.XMLGregorianCalendar?
                result_.dateTime = other.get("DateTime", "DateTime") as javax.xml.datatype.XMLGregorianCalendar?
                return result_
            } else if (other is com.gs.dmn.runtime.DMNType) {
                return toTCompositeDateTime(other.toContext())
            } else if (other is proto.TCompositeDateTime) {
                var result_: TCompositeDateTimeImpl = TCompositeDateTimeImpl()
                result_.date = com.gs.dmn.signavio.feel.lib.DefaultSignavioLib.INSTANCE.date((other as proto.TCompositeDateTime).getDate())
                result_.time = com.gs.dmn.signavio.feel.lib.DefaultSignavioLib.INSTANCE.time((other as proto.TCompositeDateTime).getTime())
                result_.dateTime = com.gs.dmn.signavio.feel.lib.DefaultSignavioLib.INSTANCE.dateAndTime((other as proto.TCompositeDateTime).getDateTime())
                return result_
            } else {
                throw com.gs.dmn.runtime.DMNRuntimeException(String.format("Cannot convert '%s' to '%s'", other.javaClass.getSimpleName(), TCompositeDateTime::class.java.getSimpleName()))
            }
        }

        @JvmStatic
        fun toProto(other: TCompositeDateTime?): proto.TCompositeDateTime {
            var result_: proto.TCompositeDateTime.Builder = proto.TCompositeDateTime.newBuilder()
            if (other != null) {
                var dateProto_: String = com.gs.dmn.signavio.feel.lib.DefaultSignavioLib.INSTANCE.string((other as TCompositeDateTime).date)
                result_.setDate(dateProto_)
                var timeProto_: String = com.gs.dmn.signavio.feel.lib.DefaultSignavioLib.INSTANCE.string((other as TCompositeDateTime).time)
                result_.setTime(timeProto_)
                var dateTimeProto_: String = com.gs.dmn.signavio.feel.lib.DefaultSignavioLib.INSTANCE.string((other as TCompositeDateTime).dateTime)
                result_.setDateTime(dateTimeProto_)
            }
            return result_.build()
        }

        @JvmStatic
        fun toProto(other: List<TCompositeDateTime?>?): List<proto.TCompositeDateTime>? {
            if (other == null) {
                return null
            } else {
                return other.stream().map({o -> toProto(o)}).collect(java.util.stream.Collectors.toList())
            }
        }
    }
}
